File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        #3 conditions
        #add ( only if the open <n
        #add ) only if the close<open
        #valid if opwn ==closed and ==n
        res = []

        def back(openP, closeP, s):
            if openP == closeP and openP + closeP == n * 2:
                res.append(s)
                return
            
            if openP < n:
                back(openP + 1, closeP, s + "(")
            
            if closeP < openP:
                back(openP, closeP + 1, s + ")")

        back(0, 0, "")

        return res

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void gen(int op,int cl,vector<string> &ans,string s,int n){
        if(op==cl&&op+cl==n*2){
            ans.push_back(s);
        }
        if(op<n){
            gen(op+1,cl,ans,s+'(',n);
        }
        if(cl<op){
            gen(op,cl+1,ans,s+')',n);
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> a;
        gen(0,0,a,"",n);
        return a;
    }
};